tests/test_charm.py

```diff
@@ -105,3 +109,107 @@ def test_new_unit_has_password(self):
             config["MYSQL_ROOT_PASSWORD"],
             self.harness.charm._stored.mysql_setup["MYSQL_ROOT_PASSWORD"],
         )
+
+    def test_provides(self):
+        with patch("charm.MySQLCharm.mysql") as mock_version:
+            VERSION = "mysql 8.0.22"
+            mock_version.return_value = VERSION
+            MySQLCharm.mysql.version = mock_version
+
+            info = {
+                "app_name": "mysql",
+                "host": "mysql-0.mysql-endpoints",
+                "port": 3306,
+                "user_name": "root",
+                "mysql_root_password": "D10S!",
+            }
+            db_info = Mock(return_value=info)
+            MySQLCharm.db_info = db_info
+
+            self.assertTrue(isinstance(self.harness.charm.provides, dict))
+            self.assertTrue("provides" in self.harness.charm.provides)
+            self.assertTrue(
+                isinstance(self.harness.charm.provides["provides"], dict)
+            )
+            self.assertEqual(
+                self.harness.charm.provides["provides"]["mysql"], VERSION
+            )
+            self.assertTrue(
+                isinstance(self.harness.charm.provides["config"], dict)
+            )
+            self.assertEqual(
+                self.harness.charm.provides["config"]["app_name"], "mysql"
```

Comment:
If you prefer, you can reduce multiple checks to a single one using `assertDictEquals`.

Reply:
Nice!
Done